### PR TITLE
Add Metadata<->Querier gRPC Stream Connection

### PR DESCRIPTION
**Pull request by nathanwilk7**
### What's being changed:

- Adds a new `QuerierStream` to the `ClusterService` cluster gRPC service which is used to establish communication between metadata<->querier nodes.
- Adds a `QuerierManager`/`Querier` to Weaviate core and uses them to send events to any connected querier nodes whenever a tenant is offloaded.
- Adds a `MetadataSubscription` to the querier nodes which registers itself with a metadata node when it receives its first search request and then downloads tenants async when they're updated in s3.
- README instructions for how to manually test the end to end flow of: write to weaviate core, offload a tenant, query that tenant on the querier, activate the tenant and insert more data, offload the tenant again, querier downloads the new tenant in the background, future queries return the newly downloaded tenant.

### Future

- Remove dependency on a single metadata node.
- Move channel out of metaclass.
- Add more tests.
- Replace grpc DialContex with NewClient (including non SS&C usage).

### Alternatives Considered

- Unidirectional gRPC stream. I originally sent un/register messages from the querier nodes to the metadata nodes, and I think it's likely we'll have other types of messages to send from querier->metadata.
- Add nonvoter querier nodes to the same raft cluster as the metadata nodes. My two concerns here are 1) that we'd need to have separate code for handling the raft log on the querier nodes vs the metadata nodes, which I expect would be confusing and high-risk to add/maintain and 2) I worry spinning up a bunch of query nodes at the same time could overwhelm the metadata nodes. Having an explicit message passing interface between the metadata<->query nodes is a nice abstraction to work with and less stressful than messing with the raft one. I expect gRPC performance to be pretty good for this type of thing as long-lived subscriptions are a fairly common use case.
- Use a message queue to send messages between metadata<->querier. Didn't want to bring in an external service like kafka/redpanda and gRPC streams were much simpler to set up than other similar things like NATS.

### Review checklist

- [x] Documentation has been updated, if necessary. Link to changed documentation: Code
- [ ] Chaos pipeline run or not necessary. Link to pipeline: N/A
- [50%] All new code is covered by tests where it is reasonable: Manual end to end test run locally, will follow up with automated tests.
- [ ] Performance tests have been run or not necessary: Not needed yet.

### CODE REVIEW

File: cluster/proto/api/message.proto [hunk 1]
```diff
@@ -2,12 +2,26 @@ syntax = "proto3";
...
   rpc RemovePeer(RemovePeerRequest) returns (RemovePeerResponse) {}
   rpc JoinPeer(JoinPeerRequest) returns (JoinPeerResponse) {}
   rpc NotifyPeer(NotifyPeerRequest) returns (NotifyPeerResponse) {}
   rpc Apply(ApplyRequest) returns (ApplyResponse) {}
   rpc Query(QueryRequest) returns (QueryResponse) {}
+
+  // QuerierStream is experimental, may be changed/removed. A QuerierStream represents a
+  // connection between this metadata cluster and a querier node. See the implementing
+  // function's doc for more information.
+  rpc QuerierStream(stream QuerierStreamRequest) returns (stream QuerierStreamResponse) {}
```
Comment (nathanwilk7) [inline cluster/proto/api/message.proto:24]: Should we move this out into a separate service? I think we'll likely modify/remove this rpc multiple times as we iterate.

File: cluster/querier/querier.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+// The notification is sent on the querier's class tenant data updates channel and is
+// non-blocking (if the channel is full, the notification is skipped).
+func (qm *QuerierManager) NotifyClassTenantDataUpdate(ct ClassTenant) error {
+	qm.mu.Lock()
+	defer qm.mu.Unlock()
+	notifyFailedErrors := []error{}
+	for q := range qm.registeredQueriers {
+		select {
+		case q.classTenantDataUpdates <- ct:
+			// TODO log debug
```
Comment (nathanwilk7) [inline cluster/querier/querier.go:61]: this and other TODOs will be addressed in follow up PRs. If any are critical to address before merging, let me know

File: exp/query/api.go [hunk 3]
```diff
@@ -112,6 +116,12 @@ func (a *API) Search(ctx context.Context, req *SearchRequest) (*SearchResponse,
 		return nil, fmt.Errorf("tenant %q is not offloaded, %w", req.Tenant, ErrInvalidTenant)
 	}
 
+	// TODO come up with a better way to ensure this doesn't run until the gRPC serving is active
```
Comment (nathanwilk7) [inline exp/query/api.go:119]: If you have any thoughts for how to solve this, let me know, this `sync.Once`/`start` stuff is obviously not how this should be

**Review by nathanwilk7 — COMMENTED** [review]
Left inline questions/comments

File: cluster/schema/meta_class.go [hunk 4]
```diff
@@ -478,6 +480,20 @@ func (m *metaClass) applyShardProcess(name string, action command.TenantProcessR
...
+			err := m.querierManager.NotifyClassTenantDataUpdate(
+				querier.ClassTenant{
+					ClassName:  m.Class.Class,
+					TenantName: name,
+				},
+			)
+			if err != nil {
+				// TODO log
+				fmt.Println("failed to notify some queriers", err)
+			}
```
Comment (nathanwilk7) [inline cluster/schema/meta_class.go:496]: @moogacs do you think this is ok to add here or should we put the call behind a config flag just in. case? The idea is that it does "nothing" if there are no querier nodes registered.

File: cluster/proto/api/message.proto (hunk 1 above)
Comment (reyreaud-l) [reply cluster/proto/api/message.proto:17]: I'd be in favor of keeping it out in a different service (maybe we make it listen on a different port as well on the weaviate nodes ? Can we have 2 service listening on the port and reuse the raft internal port ?)

Also it's a different area of responsibility `ClusterService` is related to metadata read/write and raft cluster mgmt.

@moogacs what do you think ?

File: cluster/querier/querier.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// QuerierManager keeps track of registered querier nodes and allows one to notify all of them
+// of class/tenant data updates.
+type QuerierManager struct {
+	mu                 sync.Mutex
+	registeredQueriers map[*Querier]struct{}
```
Comment (reyreaud-l) [inline cluster/querier/querier.go:25]: If you're using the lock to sync access to the map only, you can just use an `async.Map` instead to simplify the code.

File: cluster/querier/querier.go (hunk 2 above)
Comment (reyreaud-l) [reply cluster/querier/querier.go:61]: I think it would be nice for this PR to include the logger into that manager and transform println into logger call. Otherwise subsequent code might also put println and saying we'll address them in subsequent pr ;)

File: cluster/querier/querier.go (hunk 5 above)
Comment (reyreaud-l) [reply cluster/querier/querier.go:25]: ALso the sync map provides range https://pkg.go.dev/sync#Map.Range which allows us to delete a key while iterating the map elsewhere

File: cluster/rpc/server.go [hunk 6]
```diff
@@ -179,3 +184,83 @@ func toRPCError(err error) error {
...
+
+	// This channel will receive a message when a class/tenant's data has been updated
+	classTenantDataUpdates := q.ClassTenantDataUpdates()
+
+	// Start a goroutine to send class tenant data updates to the querier node.
+	//
+	enterrors.GoWrapper(func() {
+		defer wg.Done()
+		for {
+			// TODO think more about this select and stream context, do we need streamClosed?
```
Comment (reyreaud-l) [inline cluster/rpc/server.go:239]: It would be nice if that PR can include the select statement exiting on the context from the stream being closed. We forget in too many places to handle context and i'd rather we avoid adding new ones (even with a TODO).

File: cluster/rpc/server.go [hunk 7]
```diff
@@ -179,3 +184,83 @@ func toRPCError(err error) error {
...
+
+	// streamClosed should receive a message when the stream is closed by the querier node,
+	// so we can close the goroutine that sends class tenant data updates.
+	streamClosed := make(chan struct{})
+	// returnErr should be set if there is an error in the goroutines below that should
+	// be returned to the caller.
+	var returnErr error
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
```
Comment (reyreaud-l) [inline cluster/rpc/server.go:209]: I would add a comment here that add 2 because we will be waiting for 2 goroutines (one that receive messages and one that sends messages).

Also question from my end: is there a case where one goroutine will exit and not the to other ? For example the query node answer back with an error, we will exit the writing goroutine, but will we also clear the other one and boot the node ?

File: cluster/schema/meta_class.go (hunk 4 above)
Comment (reyreaud-l) [reply cluster/schema/meta_class.go:496]: Let's put it behind a config flag. The apply loop is very critical and it's better to avoid any expensive operation here. 

Could that loop just queue a message in a channel ? (with a failsafe that if the channel is full/blocking we just ignore it, log a warn and move on so that we don't bloc k?)

Then we can have one or multiple reader on that channel that will ensure the notify will be send to the queriers ? Idea being to cut out of the apply loop any code that might do a network call/hang/etc...

File: cluster/schema/meta_class.go [hunk 8]
```diff
@@ -478,6 +480,20 @@ func (m *metaClass) applyShardProcess(name string, action command.TenantProcessR
 
 		if count == len(processes) {
 			copy.Status = req.Tenant.Status
+			// Whenever a tenant is frozen, we send a notification the querier nodes registered
+			// with the querier manager. This call to NotifyClassTenantDataUpdate is non-blocking
```
Comment (reyreaud-l) [inline cluster/schema/meta_class.go:484]: `NotifyClassTenantDataUpdate` is actually blocking because it expects to acquire a lock on the querier map under the hood. See my comment below how we could make this non blocking.

**Review by reyreaud-l — COMMENTED** [review]
Great approach to the problem, I like it. The one thing I'm not sure about is the place where we wire together schema changes and shooting out the notification to the querier node.

I believe that with our current approach we might block/hang in the metaclass updates and ideally this shouldn't be possible because this code might block the raft Apply loop at a higher level. There's way we can cutout that completely (using channels and avoid notifying if the channel is full/broken) and we should use them IMO.

File: cluster/schema/meta_class.go (hunk 4 above)
Comment (nathanwilk7) [reply cluster/schema/meta_class.go:496]: > Could that operation instead just queue a message in a channel ? (with a failsafe that if the channel is full/blocking we just ignore it, log a warn and move on so that we don't bloc k?)

That is what NotifyClassTenantDataUpdate does:
```
func (qm *QuerierManager) NotifyClassTenantDataUpdate(ct ClassTenant) error {
        ....
	for q := range qm.registeredQueriers {
		select {
		case q.classTenantDataUpdates <- ct:
                        ...
		default:
                        ...
			continue
		}
	}
        ...
}
```

> Maybe we could move up the code to instead be at the raft/apply level directly and avoid the intricacies of querier notification at the schema level ? IMO that package should only care about the schema and that's it.

> Happy to discuss what is the best place to plus the notification from the schema flow (or if this location is the current best that's a-ok too!)

I'll look into these and potentially reach out with questions, thanks

File: cluster/querier/querier.go (hunk 5 above)
Comment (nathanwilk7) [reply cluster/querier/querier.go:25]: Good idea, thanks!

File: cluster/schema/meta_class.go (hunk 8 above)
Comment (nathanwilk7) [reply cluster/schema/meta_class.go:484]: Good point, I was assuming the lock would never experience contention, thanks for pointing that out

File: cluster/proto/api/message.proto (hunk 1 above)
Comment (nathanwilk7) [reply cluster/proto/api/message.proto:17]: I agree, I'll look into creating a neew grpc service/port

File: cluster/querier/querier.go (hunk 2 above)
Comment (nathanwilk7) [reply cluster/querier/querier.go:61]: Will do! I'll do as much of the "Miscellaneous logging/error handling/context/code cleanliness improvements." as I can think of

File: cluster/rpc/server.go (hunk 7 above)
Comment (nathanwilk7) [reply cluster/rpc/server.go:209]: I'll look more into this, I have handling for some basic goroutine exit scenarios but have not thought deeply about it

**Comment by kavirajk** [comment]
@nathanwilk7 @reyreaud-l I also think it's worth discussing about how it will work with multiple querier pods?

Say a tenant moves from active -> frozen, does it notify all the queriers? All querier try to download the data from local disk?

that's NxM communication and downloads for single status change.

File: cluster/schema/meta_class.go (hunk 4 above)
Comment (nathanwilk7) [reply cluster/schema/meta_class.go]: I introduced a channel layer of indirection which is only used when the metadata server is enabled. However, the only ways I could think to move the location out of the metaclass seemed overcomplicated. Let me know if you want to discuss more. I'd prefer to merge this as is given the guardrails I have in place, and revisit later.

**Comment by nathanwilk7** [comment]
> Say a tenant moves from active -> frozen, does it notify all the queriers? All querier try to download the data from local disk?
that's NxM communication and downloads for single status change.

@kavirajk  Good question, the newly updated implementation does send NxM grpc notifications, but only querier nodes which have the tenant already cached locally will download the new version. Querier nodes which receive a notification for a tenant and don't have that tenant cached will do nothing. I think this is good enough for now, in the future we can optimize this further if needed (eg by only sending version updates to nodes which have the tenant or by having the querier node require >=N time has passed since the last download for that tenant).

**Comment by nathanwilk7** [comment]
Was just thinking about this and realized that by removing the omu lock I'm not handling the case where two queries come in for the same tenant simultaneously. Will need to address this, hopefully I'll have some time later today

File: modules/offload-s3/module.go [hunk 9]
```diff
@@ -284,14 +284,18 @@ func (m *Module) Upload(ctx context.Context, className, shardName, nodeName stri
 // cloud provider (S3, Azure Blob storage, Google cloud storage)
 // {dataPath}/{className}/{shardName}/{content}
 func (m *Module) Download(ctx context.Context, className, shardName, nodeName string) error {
+	localPath := fmt.Sprintf("%s/%s/%s", m.DataPath, strings.ToLower(className), shardName)
+	return m.DownloadToPath(ctx, className, shardName, nodeName, localPath)
+}
+
+func (m *Module) DownloadToPath(ctx context.Context, className, shardName, nodeName, localPath string) error {
```
Comment (nathanwilk7) [inline modules/offload-s3/module.go:291]: @moogacs any concerns about this change?

File: cluster/schema/meta_class.go [hunk 10]
```diff
@@ -478,6 +482,25 @@ func (m *metaClass) applyShardProcess(name string, action command.TenantProcessR
 
 		if count == len(processes) {
 			copy.Status = req.Tenant.Status
+			// TODO ask loic for thoughts on how to move this out of meta class? Also does that need to be in the metaclass ?
```
Comment (nathanwilk7) [inline cluster/schema/meta_class.go:485]: @reyreaud-l or @moogacs would love your thoughts here for a follow up PR

File: exp/metadataserver/server_test.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+// TODO add tests for shutting down metadata/core/querier nodes and make sure they work as expected
+
+// testServerSetup gives you a ready to use Server and Logger and tells you which port
+// number to use.
+func testServerSetup(t *testing.T) (*Server, *logrus.Logger, int) {
+	log := logrus.New()
+	log.SetLevel(logrus.ErrorLevel)
+	// TODO replace real port with bufconn? This causes a firewall alert to trigger
```
Comment (nathanwilk7) [inline exp/metadataserver/server_test.go:36]: this and other non-critical TODOs will be addressed in follow up PRs, ran out of time to include them in this PR

File: cluster/schema/meta_class.go (hunk 10 above)
Comment (reyreaud-l) [reply cluster/schema/meta_class.go:485]: The one thing I have in mind is that we can move it to schema/manager.go -> in `UpdateTenants` we do 
```
319     return s.apply(
320         applyOp{
321             op: cmd.GetType().String(),
322             // updateSchema func will update the request's tenants and therefore we use it as a filter that is then sent
323             // to the updateStore function. This allows us to effectively use the schema update to narrow down work for
324             // the DB update.
325             updateSchema: func() error { return s.schema.updateTenants(cmd.Class, cmd.Version, req) },
326             updateStore:  func() error { return s.db.UpdateTenants(cmd.Class, req) },
327             schemaOnly:   schemaOnly,
328         },
329     )
```

We can probably add a `notifyFunc` as a parameter to `applyOp` that is used to send the tenants (in a non blocking way) to the querier nodes.

File: adapters/handlers/rest/configure_api.go [hunk 12]
```diff
@@ -374,6 +375,35 @@ func MakeAppState(ctx context.Context, options *swag.CommandLineOptionsGroup) *s
 		remoteIndexClient, appState.Logger, appState.ServerConfig.Config.Persistence.DataPath)
 	appState.Scaler = scaler
 
+	// let classTenantDataEvents be nil if the metadata server is not enabled since the metadata
+	// server/querierManager are the users of the channel
+	var classTenantDataEvents chan metadataserver.ClassTenant
+	querierManager := metadataserver.NewQuerierManager(appState.Logger)
+	metadataServer := metadataserver.NewServer(
```
Comment (reyreaud-l) [inline adapters/handlers/rest/configure_api.go:382]: Can we move that whole block of code under a config option ? Ideally I want to experimental code being executed *at all* unless we have enabled it.

File: exp/metadataserver/proto/api/metadata_server.proto [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// QuerierStreamRequest is experimental, may be changed/removed. QuerierStreamRequest is currently
+// unused as we only care about the existence of the stream to register the querier node. We'll
+// likely add new types to this message in the future and use them.
+message QuerierStreamRequest {
+  enum Type {
+    TYPE_UNSPECIFIED = 0;
+  }
+  Type type = 1;
```
Comment (reyreaud-l) [inline exp/metadataserver/proto/api/metadata_server.proto:20]: @kavirajk was raising a very good point about `NxM` broadcast. Could a follow up PR (so that we can get this one through) improve the request to have a querier node be able to specify "which" tenant/class ?

Edit: Thinking about it actually it's not ideal, cause we want to have *a* node at least have the tenant to HOT, but not all of them. And then route request to it. 

@kavirajk I think we will tackle the NxM problem once we have routing in place, we'll be able to have the metadata server manage which node has which tenant locally, and ensure that *they* receive only the status updates for these tenants. 

For now I'm fine keeping it as a broadcast, but we'll have to tackle the problem in short order, before we release it to prod.

**Review by reyreaud-l — CHANGES_REQUESTED** [review]
Hey @nathanwilk7. It's an excellent PR and I really enjoy reading the code! The comments are clear, the intent of the code is clear and it's make it easy to understand what it's doing. 

My only requested change is to tackle the "notification" part out of meta class. See my suggestion in the code, I think it can work. Happy to jump a call to discuss it once you come online.

File: modules/offload-s3/module.go (hunk 9 above)
Comment (reyreaud-l) [reply modules/offload-s3/module.go:291]: IMO It's OK

File: exp/metadataserver/proto/api/metadata_server.proto (hunk 13 above)
Comment (kavirajk) [reply exp/metadataserver/proto/api/metadata_server.proto:20]: Yep. Adding a note about the NxM problem and may be creating issue on our milestone would be nice. Feel free to skip in this PR 👍

File: exp/query/api.go [hunk 14]
```diff
@@ -315,30 +319,108 @@ func (a *API) vectorSearch(
 	return objs, distance, nil
 }
 
-func (a *API) loadOrDownloadLSM(ctx context.Context, collection, tenant string) (*lsmkv.Store, string, error) {
-	localPath := path.Join(a.offload.DataPath, strings.ToLower(collection), strings.ToLower(tenant), defaultLSMRoot)
+type LSMEnsurer interface {
```
Comment (kavirajk) [inline exp/query/api.go:322]: I would remove this interface :). 

Rationale are
1. This is used only inside this query package for now. No external package consumes it. Even any consumer wants to mock it, they can create this interface on the consumer side not on the implementation side.
2. Go interfaces are designed for [consumer side](https://go.dev/wiki/CodeReviewComments#interfaces)

To give concrete example
1. We have [SchemaQuerier](https://github.com/weaviate/weaviate/blob/main/exp/query/api.go#L77-L80) interface on `query` package [where I consume it.](https://github.com/weaviate/weaviate/blob/main/exp/query/api.go#L67)
2. But it's implemented in another package `queryschema` here. Why?
4. Because, now `query` and `queryschema` package is **not coupled**. And I link them [only](https://github.com/weaviate/weaviate/blob/main/cmd/weaviate/main.go#L62) in the [main](https://github.com/weaviate/weaviate/blob/main/cmd/weaviate/main.go#L79). If we both define and implement `SchemaQuerier` interface in `queryschema` package then we would have to use that interface in `query` package as `queryschema.SchemaQuerier` that creates coupling between `query` and `queryschema`.

In other words, it's **almost always wrong** to have both definition and implementation of interface in **same** package :). Few exceptions of course like in stdlib.

File: adapters/handlers/rest/configure_api.go (hunk 12 above)
Comment (kavirajk) [reply adapters/handlers/rest/configure_api.go:382]: Agree. +1

File: cluster/schema/manager.go [hunk 15]
```diff
@@ -38,9 +39,9 @@ type SchemaManager struct {
 	log    *logrus.Logger
 }
 
-func NewSchemaManager(nodeId string, db Indexer, parser Parser, log *logrus.Logger) *SchemaManager {
+func NewSchemaManager(nodeId string, db Indexer, parser Parser, classTenantDataEvents chan metadataserver.ClassTenant, log *logrus.Logger) *SchemaManager {
```
Comment (kavirajk) [inline cluster/schema/manager.go:42]: I would prefer to create another constructor like `NewSchemaManagerWithTenantEvents` (something like that) and use only that on our `exp/query` path.

That way we don't touch existing schemaManager and it's very explicit. Plus when not running our `exp/querier` people don't have to worry about this `classTenantDataEvents` at all. 

Wdyt? Happy to know if I'm missing any of your rationale for doing this way.

File: cluster/schema/schema.go [hunk 16]
```diff
@@ -210,11 +212,12 @@ type shardReader interface {
 	GetShardsStatus(class, tenant string) (models.ShardStatusList, error)
 }
 
-func NewSchema(nodeID string, shardReader shardReader) *schema {
+func NewSchema(nodeID string, shardReader shardReader, classTenantDataEvents chan metadataserver.ClassTenant) *schema {
```
Comment (kavirajk) [inline cluster/schema/schema.go:215]: Same here, may be another constructor something like `NewSchemaWithTenantEvents`

File: cmd/weaviate/README.md [hunk 17]
```diff
@@ -8,8 +9,13 @@ tldr; use `cmd/weaviate-server` instead. This is experimental
 
 ## Setup
 
+If you're prompted "Do you want the application “weaviate-server” to accept incoming network connections?" by your firewall select "Allow", this is because the server accepts incoming client connections (in this case from the querier).
```
Comment (kavirajk) [inline cmd/weaviate/README.md:12]: Nice. thanks for this @nathanwilk7. Can we make it as markdown NOTE type?
[some thing like this](https://github.com/orgs/community/discussions/16925)

File: cluster/schema/meta_class.go (hunk 10 above)
Comment (kavirajk) [reply cluster/schema/meta_class.go:485]: curious why not in this `meta_class.go` itself? why have to move?

File: cluster/store.go [hunk 18]
```diff
@@ -133,6 +134,11 @@ type Config struct {
 
 	EnableFQDNResolver bool
 	FQDNResolverTLD    string
+
+	// ClassTenantDataEvents can have events published onto it when tenant changes like
+	// being frozen happen, with the goal of being able to alert the metadata nodes. This
+	// channel will be nil if the metadata server is not enabled.
+	ClassTenantDataEvents chan metadataserver.ClassTenant
```
Comment (kavirajk) [inline cluster/store.go:141]: I would prefer to have some metrics about success/failure of these events. But feel free to take it in follow up PR :)

File: cmd/weaviate/README.md [hunk 19]
```diff
@@ -18,6 +24,10 @@ This also starts other dependencies like `contextionary` (for vectorizing), `min
 ### Start Weaviate Querier
 
 ``` sh
+rm -rf /tmp/question
+# Note, you may want to delete the contents of weaviate-offload minio as well, something like (add --force if you're sure):
```
Comment (kavirajk) [inline cmd/weaviate/README.md:28]: ❤️

File: cmd/weaviate/README.md [hunk 20]
```diff
@@ -252,3 +263,60 @@ with weaviate.connect_to_local(
 ```
 
 ### BM25 search (TODO)
+
+## Async Tenant Download
+
+When you freeze a tenant which is registered with the querier node, the querier node will
```
Comment (kavirajk) [inline cmd/weaviate/README.md:269]: Q: Will it work for already frozen tenants that got frozen without metadata server?

Say FROZEN (without metdata server) -> ACTIVE -> ADD more objects -> FROZEN -> Query.

Comment (kavirajk) [reply cmd/weaviate/README.md:269]: Trying to think, if we want to enable async download for already existing frozen customers on prod envs.

File: adapters/handlers/rest/configure_api.go [hunk 21]
```diff
@@ -374,6 +375,35 @@ func MakeAppState(ctx context.Context, options *swag.CommandLineOptionsGroup) *s
...
+		enterrors.GoWrapper(func() {
+			err := metadataServer.Open()
+			if err != nil {
+				appState.Logger.Errorf("metadata server did not start: %v", err)
+			}
+		}, appState.Logger)
+
+		enterrors.GoWrapper(func() {
+			for classTenantDataEvent := range classTenantDataEvents {
+				querierManager.NotifyClassTenantDataEvent(classTenantDataEvent)
```
Comment (kavirajk) [inline adapters/handlers/rest/configure_api.go:402]: curious what happens if `querierManager.NotifyClassTenantDataEvent(classTenantDataEvent)` returns error?.  Because `GoWrapper` [doesn't look like handles](https://github.com/weaviate/weaviate/blob/main/entities/errors/go_wrapper.go#L36) the error returned from `NotifyClassTenantDataEvent`. 

May be I miss something?

File: exp/metadataserver/server.go [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+		querierManager:            querierManager,
+		dataEventsChannelCapacity: dataEventsChannelCapacity,
+		log:                       log,
+	}
+}
+
+// Open starts the server and registers it as the cluster service server. Blocking.
+// Returns an error if the configured listenAddress is invalid.
+// Returns an error if the configured listenAddress is un-usable to listen on.
+func (s *Server) Open() error {
```
Comment (kavirajk) [inline exp/metadataserver/server.go:55]: Suggestion: I like long running blocking api like this to accept `ctx` and not expose separate `Close()` api.  Should respect the context cancellation and cleanup the resources if it's cancelled.

Something like

```go
svr = metadata.NewServer()
ctx, cancel = context.WithCancel()
defer cancel()

svr.Serve(ctx)
```

Rationale
1. Note. I don't need any extra go routines to start and stop. Because start() is blocking, you need another goroutine to stop it, 
2. Similar to http.ServeAndListen - Also blocking long-running api that just returns error

wdyt?

File: exp/query/api.go [hunk 23]
```diff
@@ -66,9 +68,11 @@ type API struct {
...
 	schema  SchemaQuerier
 	offload *modsloads3.Module
-	// NOTE: This lock may be in-efficient if downloaded needed from multiple tenants
-	// TODO(kavi): Optimize based on some data later.
-	omu sync.Mutex
+	// cachedTenantLsmkvStores: tenantName -> tenantLsmkvStore (lsm object and path to tenant dir on disk)
+	// the tenantLsmkvStore value is updated if new tenant data versions are downloaded.
+	// Note that tenantLsmkvStore.localTenantPath points to the root of the tenant dir,
+	// not the lsm subdir.
+	cachedTenantLsmkvStores sync.Map
```
Comment (kavirajk) [inline exp/query/api.go:75]: I would also love some metrics about this count. Just to see the memory footprint of this unbounded cardinality at scale. But something for follow up PR :)

File: exp/query/api.go [hunk 24]
```diff
@@ -315,30 +319,108 @@ func (a *API) vectorSearch(
...
+	if oldTenantLsmkvStoreAny, ok := a.cachedTenantLsmkvStores.Swap(tenant, tenantLsmkvStore{s: store, localTenantPath: localTenantTimePath}); ok {
+		// when we replace or evict an lsmkvStore, we need to shut it down.
+		// I'm assuming lsmkvStore has internal locks which prevent this shutdown call
+		// from cancelling searches running at the same time.
+		// I'm also assuming once Shutdown returns we don't need the old files on disk anymore.
+		oldTenantLsmkvStore := oldTenantLsmkvStoreAny.(tenantLsmkvStore)
+		oldTenantLsmkvStore.s.Shutdown(ctx)
+
+		// to be extra safe, only remove if the path looks reasonably like what we expect (datapath/collection/tenant/timestamp/*)
+		re := regexp.MustCompile(fmt.Sprintf(`%s/%s/\d+`, localBaseCollectionPath, tenant))
```
Comment (kavirajk) [inline exp/query/api.go:416]: regexp.Compile is expensive btw. Can we compile once as package variable and use it everywhere?

**Review by kavirajk — COMMENTED** [review]
Super nice work @nathanwilk7 ❤️ 

Overall approach sounds solid. Added few minor comments and some questions for my own understanding.

File: exp/metadataserver/proto/api/metadata_server.proto (hunk 13 above)
Comment (nathanwilk7) [reply exp/metadataserver/proto/api/metadata_server.proto:20]: added here under R&D https://github.com/weaviate/weaviate-serverless/issues/92

File: cmd/weaviate/README.md (hunk 20 above)
Comment (nathanwilk7) [reply cmd/weaviate/README.md:269]: I don't think we should automatically async download all frozen tenants as we don't currently have any indication which of them will want to be read. However, I think we may want to handle "async download for already existing frozen customers" (eg prefetching) in the future. We could do this by sending sending events or just by submitting simple fake read queries for them

File: adapters/handlers/rest/configure_api.go (hunk 21 above)
Comment (nathanwilk7) [reply adapters/handlers/rest/configure_api.go:417]: Thanks, just forgot to use the error after I added it

File: cluster/schema/meta_class.go (hunk 4 above)
Comment (nathanwilk7) [reply cluster/schema/meta_class.go]: Discussed with @reyreaud-l , given the current code structure we decided to leave this channel in meta_class.go for now and add some testing to ensure it doesn't error or block the raft apply path. Refactoring this out is probably possible, but seems like more work/risk than is worth at this time. If we stick with using the channel, we may want to revisit this in the future

File: cluster/schema/meta_class.go (hunk 10 above)
Comment (nathanwilk7) [reply cluster/schema/meta_class.go:485]: https://github.com/weaviate/weaviate/pull/5853#discussion_r1805393731

File: exp/query/api.go (hunk 14 above)
Comment (nathanwilk7) [reply exp/query/api.go]: Thanks, I got rid of the interface. I think at some point it would be nice to see if we can break out the download functionality from the API, but for now I'm just using the API as is

File: exp/metadataserver/server.go (hunk 22 above)
Comment (nathanwilk7) [reply exp/metadataserver/server.go]: Thanks, for the suggestion, I've added this. Let me know if you have feedback on the implementation, especially if you can think of any issues with how I'm using context's as it's easy to make subtle mistakes with them

File: exp/query/api.go (hunk 24 above)
Comment (nathanwilk7) [reply exp/query/api.go]: This regex is per collection/tenant so as is we can't just compile it as a package var. However, I've added a comment that if the regex is too slow we should replace it with a prefix match (dropping the timestamp check)

**Comment by nathanwilk7** [comment]
I believe the failing test `modules-acceptance-tests (--only-module-offload-s3)` is unrelated to my changes as I see the same test/error on current main: https://github.com/weaviate/weaviate/actions/runs/11384549856/job/31672505811#step:5:618

File: exp/query/config.go [hunk 25]
```diff
@@ -26,8 +26,12 @@ type Config struct {
...
 
+	// MetadataGRPCAddress is the host which will be used to connect to the metadata node's gRPC server.
+	// Note that this should be replaced later to avoid having a single metadata node as a single point of failure.
+	// If MetadataGRPCAddress is empty, the querier will connect to localhost.
+	MetadataGRPCAddress string `long:"metadata.grpc.address" description:"metadata grpc address" default:":9050"`
 	// AlwaysFetchObjectStore flag ignore what version does local querier has and fetch from
 	// object store (source of truth) all the time.
 	// NOTE: Enabling this (without any intermediate cache) can introduce more latency. Can be used to during performance testing, debugging, correctness check, etc.
-	AlwaysFetchObjectStore bool `long:"always-fetch-objectstore" description:"always fetch from object storage during query, skip local querier state." default:"false"`
+	AlwaysFetchObjectStore bool `long:"always-fetch-objectstore" description:"always fetch from object storage during query, skip local querier state."`
```
Comment (nathanwilk7) [inline exp/query/config.go:36]: when trying to run this before, i got the error:

```sh
./weaviate --target=querier
FATA[0000] failed to parse command line args             app=weaviate err="boolean flag `always-fetch-objectstore' may not have default values, they always default to `false' and can only be turned on"
```

removing the default fixed it

File: cluster/schema/meta_class_test.go [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+// Later, we should replace this test with one that uses only public types/methods.
+func TestUpdateTenantsProcess(t *testing.T) {
+	metaClassesToTest := []*metaClass{
+		{},
+		{classTenantDataEvents: make(chan metadata.ClassTenant)},
+		{classTenantDataEvents: make(chan metadata.ClassTenant, 1)},
+	}
+	for _, m := range metaClassesToTest {
+		t.Run("TestUpdateTenantsProcess", func(t *testing.T) {
+			err := addAndFreezeTenant(m)
```
Comment (reyreaud-l) [inline cluster/schema/meta_class_test.go:38]: Could you also add a case here where we go from cold to hot ?

**Review by reyreaud-l — DISMISSED** [review]
See my one comment about testing :) Great PR!

File: exp/query/config.go (hunk 25 above)
Comment (kavirajk) [reply exp/query/config.go:36]: thanks Nate.

File: usecases/config/environment.go [hunk 27]
```diff
@@ -444,6 +444,22 @@ func FromEnv(config *Config) error {
 		return fmt.Errorf("parse sentry config from env: %w", err)
 	}
 
+	config.MetadataServer.Enabled = false
+	if entcfg.Enabled(os.Getenv("EXPERIMENTAL_METADATA_SERVER_ENABLED")) {
```
Comment (kavirajk) [inline usecases/config/environment.go:448]: (not a blocker. can be a follow up PR :) )

I think better to have a cli-flag for this config as well?
`experimental.metadata-server.enabled` or something similar. 

Rationale are.
1. ENVs can easily get messy at scale (10s of ENV is fine. 100s will be crazy). I want my configs to be "local" and "explicit". ENVs are by definition "global" and "implicit".
2. I don't mind having envs, but I think we should have priorities in order,  cli-flags > envs. Basically, [this comment is no longer true unfortunately in handling configs](https://github.com/weaviate/weaviate/blob/main/usecases/config/config_handler.go#L419-L424). I would love to keep it true for things at least we add from serverless sides of things :)

**Review by kavirajk — DISMISSED** [review]
LGTM. Nice work @nathanwilk7 👍

File: cmd/weaviate/main.go [hunk 28]
```diff
@@ -110,8 +110,18 @@ func main() {
 		reflection.Register(grpcServer)
 		protocol.RegisterWeaviateServer(grpcServer, grpcQuerier)
 
+		enterrors.GoWrapper(func() {
+			metadataSubscription := query.NewMetadataSubscription(
+				a,
+				opts.Query.MetadataGRPCAddress,
+				log)
+			if err = metadataSubscription.Start(); err != nil {
```
Comment (kavirajk) [inline cmd/weaviate/main.go:118]: Q: I don't see it's stopping anywhere?

Plus I think may be it's worth accepting `ctx` and respect the cancellation, timeout, etc. given it's a blocking call?

Comment (nathanwilk7) [reply cmd/weaviate/main.go:118]: Thanks! I'll add the ctx in a follow up PR

**Review by etiennedi — APPROVED** [review]

**Comment by nathanwilk7** [comment]
test failure `[modules-acceptance-tests (--only-module-offload-s3)]` is flaky, same failure already on main: https://github.com/weaviate/weaviate/actions/runs/11384549856/job/31672505811#step:5:494 so it is unrelated to these changes, will look into flakiness next week
